### PR TITLE
Dev: utils: Check node is reachable by using both ping and ssh

### PR DESCRIPTION
**Pull request by liangxin1300**
If both ping and ssh fail, the node is considered unreachable.

To fix issue #1551

### CODE REVIEW

File: crmsh/utils.py [hunk 1]
```diff
@@ -2460,13 +2461,25 @@ def package_is_installed(pkg, remote_addr=None):
...
-    rc, _, err = ShellUtils().get_stdout_stderr("ping -c 1 {}".format(node))
-    if rc != 0:
-        raise ValueError("host \"{}\" is unreachable: {}".format(node, err))
+    rc, _, _ = ShellUtils().get_stdout_stderr(f"ping -c {ping_count} {node}")
+    if rc == 0:
+        return True
+    # ping failed, try to connect to ssh port by socket
+    error_msg = f"host \"{node}\" is unreachable"
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
```
Comment (nicholasyang2022) [inline crmsh/utils.py:2474]: Does not work with IPv6. Please use `getaddrinfo` first.

File: crmsh/utils.py [hunk 2]
```diff
@@ -2460,13 +2461,25 @@ def package_is_installed(pkg, remote_addr=None):
...
-def ping_node(node):
+def node_reachable_checking(node, ping_count=1, socket_port=22, socket_timeout=3):
     """
-    Check if the remote node is reachable
+    Check if node is reachable by using ping and socket to ssh port
     """
-    rc, _, err = ShellUtils().get_stdout_stderr("ping -c 1 {}".format(node))
-    if rc != 0:
-        raise ValueError("host \"{}\" is unreachable: {}".format(node, err))
+    rc, _, _ = ShellUtils().get_stdout_stderr(f"ping -c {ping_count} {node}")
```
Comment (nicholasyang2022) [inline crmsh/utils.py:2468]: Please add `-n` to avoid triggering a reverse DNS lookup.

File: crmsh/utils.py (hunk 1 above)
Comment (liangxin1300) [reply crmsh/utils.py:2474]: I found we already have `utils.check_port_open`, should be reused this time

File: crmsh/utils.py [hunk 3]
```diff
@@ -2460,13 +2467,18 @@ def package_is_installed(pkg, remote_addr=None):
...
-def ping_node(node):
+def node_reachable_checking(node, ping_count=1, port=22, timeout=3):
     """
-    Check if the remote node is reachable
+    Check if node is reachable by using ping and socket to ssh port
     """
-    rc, _, err = ShellUtils().get_stdout_stderr("ping -c 1 {}".format(node))
-    if rc != 0:
-        raise ValueError("host \"{}\" is unreachable: {}".format(node, err))
+    rc, _, _ = ShellUtils().get_stdout_stderr(f"ping -n -c {ping_count} {node}")
```
Comment (nicholasyang2022) [inline crmsh/utils.py:2474]: `timeout` is not applied to the ping command.

File: crmsh/utils.py [hunk 4]
```diff
@@ -2460,13 +2467,18 @@ def package_is_installed(pkg, remote_addr=None):
     return rc == 0
 
 
-def ping_node(node):
+def node_reachable_checking(node, ping_count=1, port=22, timeout=3):
```
Comment (nicholasyang2022) [inline crmsh/utils.py:2470]: "check" is a noun.

```suggestion
def node_reachable_check(node, ping_count=1, port=22, timeout=3):
```

**Review by nicholasyang2022 — APPROVED** [review]

**Pull request merged**